File: src/main/java/hu/kirdev/schpincer/web/MainController.java
package hu.kirdev.schpincer.web;

import hu.kirdev.schpincer.model.*;
import hu.kirdev.schpincer.service.*;
import io.swagger.annotations.Api;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;

import javax.servlet.http.HttpServletRequest;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Controller
public class MainController {

    @Autowired
    CircleService circles;
    
    @Autowired
    OpeningService openings;
    
    @Autowired
    ItemService items;
    
    @Autowired
    ReviewService reviews;
    
    @Autowired
    OrderService orders;

    @Autowired
    ControllerUtil util;
    
    @GetMapping("/")
    public String root(HttpServletRequest request, Map<String, Object> model) {
        List<CircleEntity> circlesList = circles.findAllForMenu();
        model.put("circles", circlesList);
        List<CircleEntity> random = new ArrayList<>(circlesList);
        Collections.shuffle(random);
        model.put("circlesRandom", random);
        List<OpeningEntity> opens = openings.findUpcomingOpenings();
        model.put("opener", opens.size() > 0 ? opens.get(0) : null);
        model.put("openings", openings.findNextWeek());

        UserEntity user = util.getUser(request);
        model.put("orders", Collections.EMPTY_LIST);
        if (user != null) {
            model.put("orders", this.orders.findAll(user.getUid()).stream()
                    .filter(x -> x.getStatus() == OrderStatus.ACCEPTED)
                    .limit(3)
                    .collect(Collectors.toList()));
        }

        return "index";
    }
    
    @GetMapping("/items")
    public String items(@RequestParam(name = "q", defaultValue = "", required = false) String keyword, 
    		Map<String, Object> model) {

        model.put("circles", circles.findAllForMenu());
        model.put("searchMode", !("".equals(keyword)));
        model.put("keyword", keyword);
        return "items";
    }

    @GetMapping("/szor")
    public String circle(Map<String, Object> model) {
        model.put("circles", circles.findAllForMenu());
        model.put("circlesWithOpening", circles.findAllForInfo());
        return "circle";
    }
    
    @GetMapping("/circle/{circle}")
    public String circleSpecific(@PathVariable String circle, 
            Map<String, Object> model) {
        
        model.put("circles", circles.findAllForMenu());

        if (circle.matches("^\\d+$")) {
            long id = Long.parseLong(circle);
            model.put("selectedCircle", circles.getOne(id));
            model.put("nextOpening", openings.findNextStartDateOf(id));
        } else {
            CircleEntity circleEntity = circles.findByAlias(circle);
            model.put("selectedCircle", circleEntity);
            model.put("nextOpening", openings.findNextStartDateOf(circleEntity.getId()));
        }
        return "circleProfile";
    }

    @GetMapping({"/provider/{circle}", "/p/{circle}"})
    public String circleSpecificAlias(@PathVariable String circle, Map<String, Object> model) {
        return circleSpecific(circle, model);
    }

    @GetMapping("/profile")
    public String profile(HttpServletRequest request, Map<String, Object> model) {
        List<OrderEntity> orders = this.orders.findAll(util.getUser(request).getUid());
        Collections.reverse(orders);

        model.put("orders", orders);
        model.put("circles", circles.findAllForMenu());
        return "profile";
    }
    
}

File: src/main/kotlin/hu/kirdev/schpincer/web/MainController.kt
package hu.kirdev.schpincer.web

import hu.kirdev.schpincer.model.CardType
import hu.kirdev.schpincer.model.CircleEntity
import hu.kirdev.schpincer.model.OrderStatus
import hu.kirdev.schpincer.service.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import org.springframework.security.core.Authentication
import java.time.Instant

@Controller
open class MainController {

    @Autowired
    private lateinit var circles: CircleService

    @Autowired
    private lateinit var openings: OpeningService

    @Autowired
    private lateinit var orders: OrderService

    @Autowired
    private lateinit var timeService: TimeService

    @Autowired
    private lateinit var statService: StatisticsService

    @Autowired
    private lateinit var config: RealtimeConfigService

    @GetMapping("/")
    fun root(auth: Authentication?, model: Model, @RequestParam(defaultValue = "") error: String): String {
        val circlesList: List<CircleEntity> = circles.findAllForMenu()
        model.addAttribute("circles", circlesList)

        val random = ArrayList(circlesList)
        random.shuffle()
        model.addAttribute("circlesRandom", random)

        val opens = openings.findUpcomingOpenings()
        model.addAttribute("opener", if (opens.isNotEmpty()) opens[0] else null)
        model.addAttribute("openings", openings.findNextWeek())
        model.addAttribute("error", error)

        model.addAttribute("orders", Collections.EMPTY_LIST)
        if (auth.hasUser()) {
            model.addAttribute("orders", orders.findAll(auth.getUserId()!!)
                    .filter { it.status === OrderStatus.ACCEPTED }
                    .filter { it.date >= (Instant.now().toEpochMilli() - (1000 * 60 * 60 * 24 * 7 * 3)) }
                    .take(3))
        }
        model.addAttribute("timeService", timeService)
        config.injectPublicValues(model)
        return "index"
    }

    @GetMapping("/items")
    fun items(@RequestParam(name = "q", defaultValue = "") keyword: String, model: Model, auth: Authentication?): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("searchMode", "" != keyword)
        model.addAttribute("keyword", keyword)
        model.addAttribute("card", (auth.getUserIfPresent()?.cardType ?: CardType.DO).name)
        config.injectPublicValues(model)
        return "items"
    }

    @GetMapping("/falatozo")
    fun circle(model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("circlesWithOpening", circles.findAllForInfo())
        model.addAttribute("timeService", timeService)
        config.injectPublicValues(model)
        return "circle"
    }

    @GetMapping("/circle/{circle}")
    fun circleSpecific(@PathVariable circle: String, model: Model, auth: Authentication?): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("card", (auth.getUserIfPresent()?.cardType ?: CardType.DO).name)
        if (circle.matches("^\\d+$".toRegex())) {
            val id = circle.toLong()
            model.addAttribute("selectedCircle", circles.getOne(id))
            model.addAttribute("nextOpening", openings.findNextStartDateOf(id))
        } else {
            val circleEntity: CircleEntity = circles.findByAlias(circle)
            model.addAttribute("selectedCircle", circleEntity)
            model.addAttribute("nextOpening", openings.findNextStartDateOf(circleEntity.id))
        }
        model.addAttribute("timeService", timeService)
        config.injectPublicValues(model)
        return "circleProfile"
    }

    @GetMapping(path = ["/provider/{circle}", "/p/{circle}"])
    fun circleSpecificAlias(@PathVariable circle: String, model: Model, auth: Authentication?): String {
        return circleSpecific(circle, model, auth)
    }

    data class OrderForDetails(
        val id: Long,
        val name: String,
        val count: Int,
        val price: Int,
        val image: String,
        val color: String,
        val room: String,
        val comment: String,
        val changable: Boolean
    )

    @GetMapping("/profile")
    fun profile(auth: Authentication?, model: Model): String {
        val orders = this.orders.findAll(auth.getUserId()!!)
        model.addAttribute("orders", orders)
        model.addAttribute("ordersForDetails", orders.map {
            OrderForDetails(
                it.id, it.name, it.count, it.price,
                it.orderedItem?.imageName ?: "/image/blank-item.jpg",
                it.orderedItem?.circle?.cssClassName ?: "blank",
                it.room,
                it.comment.replace(Regex("^\\[((AB)|(KB)|(DO))] "), ""),
                it.cancelUntil > Instant.now().toEpochMilli() && it.status == OrderStatus.ACCEPTED
            )
        })
        model.addAttribute("priceBreakdowns", this.orders.generatePriceBreakdowns(orders))
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("timeService", timeService)
        model.addAttribute("uid", auth.getUserId()!!.sha256().substring(0, 6))
        config.injectPublicValues(model)
        return "profile"
    }

    private val statsViews: ConcurrentHashMap<String, String> = ConcurrentHashMap<String, String>()

    @GetMapping("/stats")
    fun stats(auth: Authentication?, model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        val user = auth.getUser()
        statsViews.computeIfPresent(user.uid) { _, b -> "$b+1" }
        statsViews.computeIfAbsent(user.uid) { user.name + ";" + Instant.now().toEpochMilli() + ";1" }
        statService.getDetailsForUser(user).entries.forEach { model.addAttribute(it.key, it.value) }
        config.injectPublicValues(model)
        return "stats"
    }

    @GetMapping("/admin/stats-insight")
    @ResponseBody
    fun statsInsights(auth: Authentication?): String {
        return if (auth.getUserIfPresent()?.sysadmin == true) {
            statsViews.values.toString()
        } else {
            "Nice try!"
        }
    }

}
